### ISSUE TITLE
taint fields: timeAdded not set

### ISSUE THREAD

**Issue by justinsb** [kubernetes/kubernetes#42394]
I set taints on a node using the kubelet, but the timeAdded field was not set:

```
 taints:
  - effect: NoSchedule
    key: dedicated
    timeAdded: null
    value: master
```

In addition, I think we prefer to omit null fields.

**Comment by liggitt** [comment]
doc for that field says it is only written for NoExecute taints, and that field is omitempty

**Comment by liggitt** [comment]
ah, looks like omitempty will only work if timeAdded is a pointer

**Comment by liggitt** [comment]
opened speculative PR in https://github.com/kubernetes/kubernetes/pull/43016 to change it to a pointer (like other optional time fields like deletionTimestamp)

**Comment by liggitt** [comment]
@davidopp what component is supposed to set timeAdded for NoExecute taints (and is validation supposed to prevent it for taints with effects other than NoExecute)? The only place I see it getting set is by the node controller

**Comment by davidopp** [comment]
As @liggitt alluded to, strictly speaking timeAdded is only needed for taints of effect NoExecute. The field is needed in order to make tolerationSeconds work in the face of component failure  -- whichever component enforces the taint will evict the pod if timeAdded + pod.tolerationSeconds <= time.Now. (Yes, clock skew will break this if they are different components.) So only the components that add NoExecute taints (which is only NodeController) set timeAdded, and only for NoExecute taints.

#40586 was filed with the suggestion to add it for all taints. It's not strictly necessary, but since the field is user-visible (despite being purely an "internal" concept), it would probably be less confusing if we were consistent and set timeAdded for all taint types.

For 1.6 I think the current behavior is fine.

These comments are unrelated to #43016, they are just about whether/when we should be setting timeAdded.

Let me know if this is unclear.

**Comment by aveshagarwal** [comment]
Other components where taints can be added are: kubelet (-register-with-taints) and kubectl taints. Right now, taint controller enforces pod eviction for NoExecute taints.

**Comment by aveshagarwal** [comment]
Another component is node controller that adds taints for node problems (unreachable, notready) but clock skew is not an issue here, but only for kubelet, kubectl taint I think.

**Comment by liggitt** [comment]
If the switch to a pointer won't change the values servers will accept, then current behavior seems ok for 1.6 (for JSON, we'll need to handle `null`, so that'll be fine... I'm less sure about protobuf)

**Comment by ethernetdan** [comment]
@davidopp @aveshagarwal @liggitt should this be blocking for 1.6?

**Comment by davidopp** [comment]
There are two issues here. One is just a dup of #40586. The main question is whether #43016 is blocking for 1.6. I don't know enough about the API stuff to know for sure.

**Comment by liggitt** [comment]
@smarterclayton says proto always tolerates missing objects, so I think #43016 is ok to move to 1.7

**Comment by k82cn** [comment]
The PR was moved to 1.8, should we also move this issue to 1.8?

**Comment by spiffxp** [comment]
friendly v1.8 release ping, does this issue need to be in the v1.8 milestone? is this issue actively being worked?

it looks like #43016 which partially fixes this issue has been moved to v1.9

**Comment by liggitt** [comment]
no, moving to 1.9

**Comment by k82cn** [comment]
/status approved-for-milestone

We'd like to get this fixed in 1.9.

**Comment by dims** [comment]
@k82cn @justinsb move to 1.10?

**Comment by k82cn** [comment]
Fixed by #43016.
